finality-aleph/src/party.rs

```diff
@@ -244,126 +260,52 @@ where
+        let client = self.client.clone();
+        let finalization_proposals_rx = &mut self.finalization_proposals_rx;
+        let proposals_task = {
+            let client = self.client.clone();
+            async move {
+                while client.info().finalized_number < current_stop_h {
+                    if let Some(proposal) = finalization_proposals_rx.next().await {
+                        // TODO: check if we should do this
+                        finalize_block(
+                            client.clone(),
+                            proposal.hash,
+                            proposal.number,
+                            Some((ALEPH_ENGINE_ID, proposal.justification)),
+                        );
+                    } else {
+                        debug!(target: "afa", "the channel of proposed blocks closed unexpectedly");
                     }
                 }
             }
+        };
 
-            if let Some(instance) = self.sessions.remove(&curr_id) {
-                if let Some(exit_tx) = instance.exit_tx {
-                    // Signal the end of the session
-                    debug!(target: "afa", "Signaling end of the consensus party #{}.", curr_id);
-                    exit_tx.send(()).expect("Closing member session");
-                }
-
-                self.sessions.insert(
-                    curr_id,
-                    SessionInstance {
-                        session: instance.session,
-                        exit_tx: None,
-                    },
-                );
-            }
-
-            debug!(target: "afa", "Moving to new session #{}.", curr_id + 1);
-
-            // We are asking for the next_session on the last block, which we know that is finalized, of the current session.
-            // We are calling the `current_session` function not the `next_session` the sessions rotates on the last blocks of the session.
-            let next_session = match self
-                .client
-                .runtime_api()
-                .current_session(&BlockId::Number(current_stop_h))
-            {
-                Ok(session) => session,
-                _ => {
-                    error!(target: "afa", "No next session found for current block #{}", current_stop_h);
-                    return;
-                }
-            };
-            // Start new session if we are in the authority set.
-            let (instance, proposition_rx) = create_session(
-                self.authority.clone(),
-                self.auth_keystore.clone(),
-                self.select_chain.clone(),
-                self.spawn_handle.clone(),
-                next_session,
-                &session_manager,
-            );
-            self.sessions.insert(curr_id + 1, instance);
-            if let Some(proposition_rx) = proposition_rx {
-                proposition_select.push(futures::stream::repeat(curr_id + 1).zip(proposition_rx));
-            }
+        tokio::select! {
```

Comment:
Seems like we are selecting out of one option here

Reply:
fixed

Reply:
Also, there was a problem about crush tolerance, currently we poll the finalization proposals concurrently with running sessions, so in the case we join a session which is already done, we will finish once the finalization proofs arrive.